RAxML support for Galaxy

tiagoantao opened the pull request:
As per the subject, this adds RAxML support for Galaxy. Based on original work by @moskalenko, which I updated a little bit.

bgruening commented:
@tiagoantao awesome to see you here! :)

bgruening commented on tools/raxml/raxml.xml:
```diff
@@ -0,0 +1,564 @@
+            </param>
+        </when>
+    </conditional>
+    <param name="random_seed" type="integer" value="1234567890" size="12" label="Random
```
You could make this optional and use `$RANDOM` in you cheetah code to pick a random integer during every run.
Not sure if `$RANDOM` works under OS-X.

nsoranzo replied:
@bgruening But having this option is good for reproducibility.

bgruening replied:
Ok, maybe this was misleading. This cheetah code does not really fit the param here:
```
    #if $random_seed:
        --random_seed $random_seed
    #else
        --random_seed 1234567890
    #end if
```
So what should be the default behaviour? I guess a fixed integer. 
But if the user wants to generate a random new number, this should be possible. So maybe replace this param with a boolean asking to generate a random number?
Nevermind, this is too nitpicking. I was just stumbling about it.

@tiagoantao can you remove all `size=""` attributes as well? This is not needed anymore.

tiagoantao replied:
I have removed the sizes.

bgruening commented on tools/raxml/raxml.xml:
```diff
@@ -0,0 +1,564 @@
+            </param>
+            ## Alphabetical Listing of Advanced Options
+            ## (-a)
+            <param format="txt" name="weightfile" type="data" label="Column weight file" optional="True" />
```
Column weight file, sounds like a tabular?

tiagoantao replied:
Documentation suggests that any white space is acceptable. What would be your advice in that case. Still tabular?

bgruening replied:
In this case txt is a good choice. It will also display tabular files.

bgruening commented on tools/raxml/raxml.xml:
```diff
@@ -0,0 +1,564 @@
+  <outputs>
+        <data format="txt" name="info" from_work_dir="RAxML_info.galaxy" label="Info" />
+      ## REQUIRED
+        <data format="txt" name="logReq" from_work_dir="RAxML_log.galaxy" label="Log">
```
Is this really all text output. Galaxy has a lot phylogenetic datatypes which you could reuse.

tiagoantao replied:
Yep, nhx for quite a few

bgruening commented on tools/raxml/raxml.xml:
```diff
@@ -0,0 +1,564 @@
+Randomized Axelerated Maximum Likelihood
+
+Author:
+_______
```
Not sure this is needed, as we have the citation. At least this is unusual :)

tiagoantao replied:
Sorry, what precisely?

bgruening replied:
The author section. A classical help section looks usually like this: https://github.com/galaxyproject/tools-iuc/blob/master/tools/pear/pear.xml

bgruening commented on tools/raxml/raxml.xml:
```diff
@@ -0,0 +1,564 @@
+extremely large datasets, either in terms of the number of taxa and/or the
+sequence length.
+
+This wrapper works with the RAxML version 8.2.4.
```
This line can be remove, as this information is provided by Galaxy in a more dynamic way.

bgruening commented on tools/raxml/.shed.yml:
```diff
@@ -0,0 +1,6 @@
+name: raxml
+owner: tiagoantao
```
iuc

bgruening commented on packages/package_raxml_8_2_4/tool_dependencies.xml:
```diff
@@ -0,0 +1,23 @@
+        <action type="shell_command">make -f Makefile.PTHREADS.gcc</action>
+
+        <!-- Install raxml -->
+        <action type="shell_command">mkdir -p $INSTALL_DIR/bin</action>
```
Any reason you need the extra directory ./bin here?

bgruening commented on packages/package_raxml_8_2_4/.shed.yml:
```diff
@@ -0,0 +1,6 @@
+description: Tool dependency for RAxML - A Maximum Likelihood based phylogenetic inference
+type: tool_dependency_definition
+categories:
+- Phylogenetics
```
`Tool Dependency Packages`
If you like you can add a long description:
```
RAxML   (Randomized   Axelerated   Maximum   Likelihood)   is   a   program   for   sequential   and   parallel
Maximum   Likelihood   based   inference   of   large   phylogenetic   trees.   It   can   also   be   used   for   post-
analyses of sets of phylogenetic trees, analyses of alignments and, evolutionary placement of short
reads.
http://sco.h-its.org/exelixis/web/software/raxml/index.html
```

bgruening commented on packages/package_raxml_8_2_4/.shed.yml:
```diff
@@ -0,0 +1,6 @@
+name: package_raxml_8_2_4
+owner: tiagoantao
```
iuc

bgruening commented on tools/raxml/README.md:
```diff
@@ -0,0 +1,10 @@
+RAxML for Galaxy
+================
```
Could you rename this one to acknowledgement.md or add some info about the tools to this one. Moreover, if you would like to display this one on the TS, you should make a rst file out of it.

bgruening commented on tools/raxml/raxml.xml:
```diff
@@ -0,0 +1,564 @@
+                <has_text_matching expression="Frog" />
+              </assert_contents>
+          </output>
+      </test>
```
You are using planemo right? Would you mind to add a few more tests to check for some of the additional outputs?

bgruening commented on tools/raxml/raxml.xml:
```diff
@@ -0,0 +1,564 @@
+Oleksandr Moskalenko with adaptations from Tiago Antao.
+
+  </help>
+<citations>
```
indentation ...

bgruening commented on tools/raxml/tool_dependencies.xml:
```diff
@@ -0,0 +1,6 @@
+<?xml version="1.0"?>
+<tool_dependency>
+  <package name="raxml" version="8.2.4">
+    <repository name="package_raxml_8_2_4" owner="tiagoantao" toolshed="https://testtoolshed.g2.bx.psu.edu" />
```
iuc
and the toolshed attribute should be removed

bgruening commented on tools/raxml/raxml.xml:
```diff
@@ -0,0 +1,564 @@
+  </requirements>
+  <command interpreter="python">raxml.py
+    ## Required parameters 
+    ## Program name and threads are hard-coded as they are dependent on the binary and threads are specified in the Galaxy's tool configuration
```
This text needs some adoption ..

tiagoantao replied:
adoption????

martenson replied:
adaptation probably

bgruening replied:
Upps, yes adaptation ... This comment is not completely true anymore, isn't?

tiagoantao commented:
I hope to have addressed all comments save two:

1. The Random issue
2. More tests (will work on this later today).

If there are other issues pending, give me a shout...

bgruening commented:
@tiagoantao thanks. Please ping me if you have pushed the changes.

tiagoantao commented:
@bgruening I am slightly desperate with one of my tests that seems to be failing. The file exists when I do a planemo shed_test with --no_cleanup and has the relevant content. I am going slight mad here. I suspect I am doing a basic mistake, but I do not know where...

bgruening commented:
@tiagoantao I will have a look, which test in particular?

tiagoantao commented:
@bgruening The second one... (for now there are only two).
Thanks!

bgruening commented on packages/package_raxml_8_2_4/tool_dependencies.xml:
```diff
@@ -0,0 +1,23 @@
+        <action type="shell_command">make -f Makefile.PTHREADS.gcc</action>
+
+        <!-- Install raxml -->
+        <action type="shell_command">mkdir -p $INSTALL_DIR</action>
```
You don't need to create the dir.

bgruening commented on packages/package_raxml_8_2_4/tool_dependencies.xml:
```diff
@@ -0,0 +1,23 @@
+
+        <!-- Install raxml -->
+        <action type="shell_command">mkdir -p $INSTALL_DIR</action>
+        <action type="shell_command">cp raxmlHPC-PTHREADS $INSTALL_DIR</action>
```
you can use the move_file action-type

bgruening commented on tools/raxml/raxml.xml:
```diff
@@ -0,0 +1,573 @@
+          <param name="model_type" value="nucleotide"/>
+          <param name="base_model" value="GTRCAT"/>
+          <param name="number_of_runs" value="5"/>
+          <output name="parsimonyTree" file="RAxML_parsimonyTree.galaxy">
```
@tiagoantao this file here is the same as from test one. Both tests generate different output, but you are comparing it with the same file this will fail.

bgruening commented:
Please, let me know if this does help. You can simply change the name and generate new test data with `--update_test_data` in planemo.

tiagoantao commented:
@bgruening after much planemo suffering, here is my suggestion. Tell me if you want another change(s)...

martenson commented:
@tiagoantao what trouble did you have with Planemo?

tiagoantao commented:
@martenson A lot of stuff. Problems from 0.19 to 0.20. Documentation not clear. Strange test workflow. If I were more mature with galaxy internals I would volunteer to help with changing it. As it is, planemo is the only thing I particularly dislike within galaxy.

As I get more mature developing tools, I hope to do more constructive comments than this...

martenson commented:
@tiagoantao I am sorry to hear that :( Could you please point me to the parts fo the documentation that were unclear?

tiagoantao commented:
@martenson , I had several problems. I still have to be quite frank. Note that some problems below are not documentation related:

Understanding the correct workflow. When should I use test, when should I use shed_test?
What are the parameters for testing? The documentation (which is just a pull of the --help output) is not very clear about that. Sometimes I did not know if I was testing all the tools in a directory if or a single tool (this was particularly bad when I started working on a fork of tools-iuc).

The speed of shed_test (with all the overheads)

The output of the tests. For example, some cases the XML files were not correct, no help there. Also the format of the output on HTML is not helpful.

Suggestions:
Put some concrete examples of usage in the documentation. Examples with shed_test and test.
Make suggestions on planemo workflows.
Run planemo lint before testing (automatically). I spent hours testing things that I should have linted. While this was a mistake of mine, the tool could actually have made life easier here.
Easier test output to interpret...

martenson commented:
@tiagoantao I created an issue for it: https://github.com/galaxyproject/planemo/issues/382

bgruening commented on tools/raxml/.shed.yml:
```diff
@@ -0,0 +1,6 @@
+owner: iuc
+description: RAxML - A Maximum Likelihood based phylogenetic inference
+type: unrestricted
+categories:
```
Note for myself: I will update this file post merge

bgruening commented on tools/raxml/raxml.xml:
```diff
@@ -0,0 +1,593 @@
+    ## (-T)
+    ## Cannot have --threads 1
+    #set $slots = $getVar('GALAXY_SLOTS', 1)
+    #if $slots == 1:
```
This is nice!

bgruening commented on tools/raxml/raxml.xml:
```diff
@@ -0,0 +1,593 @@
+                <option value="S16B">S16B</option>
+            </param>
+            ## (-b)
+            <param name="bootseed" type="integer" value="" label="Random number for non-parametric bootstrapping" refresh_on_change='true'  optional="True" />
```
refresh_on_change? Is this really needed?

bgruening commented on tools/raxml/raxml.xml:
```diff
@@ -0,0 +1,593 @@
+            ## (-x)
+            <param name="rapid_bootstrap_random_seed" type="integer" value='' label="Rapid bootstrapping random seed" optional="True" help="Specify a random seed and turn on rapid bootstrapping. CAUTION: unlike in version 7.0.4 RAxML will conduct rapid BS replicates under the model of rate heterogeneity you specified via '-m' and not by default under CAT." />
+            ## (-B)
+            <param name="cutoff_threshold" type="float" label="MR Cutoff threshold" value="" help="Cutoff threshold for the MR-based bootstopping criteria, recommended value is 0.1" optional="True" />
```
@tiagoantao last big change I would like to see (but I can also do this post-merge if you like) is to add the original parameter (-B, -c) as part of the help="" text. Usually we do things like this `help="foo bar. (-B)"`

bgruening commented on tools/raxml/raxml.xml:
```diff
@@ -0,0 +1,593 @@
+            ## (-d)
+            <param name="search_complete_random_tree" type="boolean" checked="False" truevalue="--search_complete_random_tree" falsevalue="" display="checkboxes" label="Start ML optimization from a complete random starting tree" />
+            ## (-D)
+            <param name="ml_search_convergence" type="boolean" checked="False" truevalue="--ml_search_convergence" falsevalue="" label="ML search convergence criterion" />
```
FYI: Galaxy has a agument="" attribute that can be used in this situation. If you write argument="--ml_search_convergence" the default `name` will be `ml_search_convergence` without the need to write this down and `--ml_search_convergence` is automatically added to your help text.

tiagoantao replied:
Arghh... changed the help before reading this...

bgruening commented on tools/raxml/raxml.xml:
```diff
@@ -0,0 +1,593 @@
+            <filter>(selExtraOpts['extraOptions'] == "full")</filter>
+	    <!--
+            <filter>(selExtraOpts['rapid_bootstrap_random_seed'] == '')</filter>
+            <filter>(selExtraOpts['bootseed'] == '')</filter> 
```
Is this needed?

bgruening commented on tools/raxml/tool_dependencies.xml:
```diff
@@ -0,0 +1,6 @@
+<?xml version="1.0"?>
+<tool_dependency>
+  <package name="raxml" version="8.2.4">
+    <repository name="package_raxml_8_2_4" owner="tiagoantao" />
```
iuc?

bgruening commented:
@tiagoantao this wrapper is great! Thanks you very much to contribute this to IUC and sorry for my constant nitpicking. I'm also sorry for you bad experiences you had with planemo. I actually love planemo but you are right, maybe it is to developer centric. I sure we will be able to make it much more pleasant until you will work on RadSeq :)

Thanks again!
Bjoern

tiagoantao commented:
@bgruening Thanks for the comments. I have updated this. The command line parameters are on the help...

bgruening commented:
@erasche @nsoranzo can we merge this? I have a few small style things I would like to update post-merge, but I have annoyed @tiagoantao too much already. Is there any substantial bug that we have overseen?

nsoranzo commented on packages/package_raxml_8_2_4/tool_dependencies.xml:
```diff
@@ -0,0 +1,30 @@
+    <install version="1.0">
+      <actions>
+        <!-- Download source code -->
+        <action type="download_by_url" target_filename="standard-RAxML-8.2.4.tar.gz">https://github.com/stamatak/standard-RAxML/archive/v8.2.4.tar.gz</action>
```
Can you add the `sha256sum` attribute?

nsoranzo commented on packages/package_raxml_8_2_4/tool_dependencies.xml:
```diff
@@ -0,0 +1,30 @@
+        <!-- Install raxml -->
+        <action type="move_file">
+          <source>raxmlHPC-PTHREADS</source>
+	  <destination>$INSTALL_DIR/bin</destination>
```
Indentation.

nsoranzo commented on packages/package_raxml_8_2_4/tool_dependencies.xml:
```diff
@@ -0,0 +1,30 @@
+        </action>
+        <action type="move_file">
+          <source>raxmlHPC</source>
+	  <destination>$INSTALL_DIR/bin</destination>
```
Indentation.

martenson commented:
This PR seems solid to me. :+1: 

@nsoranzo I will do the stylistic changes you pointed out post-merge.

nsoranzo commented on tools/raxml/raxml.xml:
```diff
@@ -0,0 +1,593 @@
+  pages={1312--1313},
+  year={2014},
+  publisher={Oxford Univ Press}}
+    </citation>
```
This will add the same citation twice! You can leave just the one with the `doi`.

bgruening commented:
@martenson thanks for offering, I also think we should fix (small corrections) post-merge to not entirely demotivate @tiagoantao :)

@nsoranzo thanks for reviewing!

nsoranzo commented on tools/raxml/raxml.xml:
```diff
@@ -0,0 +1,593 @@
+            <param name="print_branch_lengths" type="boolean" checked="no" truevalue="--print_branch_lengths" falsevalue="" display="checkboxes" label="Print bootstrapped trees with branch lengths" help="Specifies that bootstrapped trees should be printed with branch lengths. The bootstraps will run a bit longer, because model parameters will be optimized at the end of each run under GAMMA or GAM ...
+            ## (-K)
+            <param name="multistate_sub_model" type="select" label="Specify a
+                multi-state substitution model" optional="True">
```
Add `help="(-K)"`

nsoranzo commented on tools/raxml/raxml.xml:
```diff
@@ -0,0 +1,593 @@
+
+    </command>
+  <inputs>
+    ## (-s)
```
`##` starts a comment only in Cheetah, for XML you should use `<!-- ... -->` .

Anyway these can all probably be removed now that the parameter switches have been documented in the `help` attribute.

nsoranzo commented on tools/raxml/raxml.xml:
```diff
@@ -0,0 +1,593 @@
+    ## (-n) - see elsewhere
+    ## (-o)
+    #if $selExtraOpts.outgroup_name:
+        --outgroup_name $selExtraOpts.outgroup_name
```
`$selExtraOpts.outgroup_name` must be quoted here!

nsoranzo commented on tools/raxml/raxml.xml:
```diff
@@ -0,0 +1,593 @@
+        --threads $slots
+    #end if
+    ## (-s)
+    --source $infile
```
Quote `$infile` because param of type `data` can contain spaces.

nsoranzo commented on tools/raxml/raxml.xml:
```diff
@@ -0,0 +1,593 @@
+    #end if
+    ## (-a)
+    #if $selExtraOpts.weightfile:
+        --weightfile $selExtraOpts.weightfile
```
Quote `$selExtraOpts.weightfile`

nsoranzo commented on tools/raxml/raxml.xml:
```diff
@@ -0,0 +1,593 @@
+    #end if
+    ## (-E)
+    #if $selExtraOpts.excludefile:
+        --excludefile $selExtraOpts.excludefile
```
Quote `$selExtraOpts.excludefile`

nsoranzo commented on tools/raxml/raxml.xml:
```diff
@@ -0,0 +1,593 @@
+    $selExtraOpts.save_memory_cat_model
+    ## (-g)
+    #if $selExtraOpts.groupingfile:
+        --groupingfile $selExtraOpts.groupingfile
```
Quote `$selExtraOpts.groupingfile`

nsoranzo commented on tools/raxml/raxml.xml:
```diff
@@ -0,0 +1,593 @@
+    $selExtraOpts.disable_undetermined_seq_check
+    ## (-P)
+    #if $selExtraOpts.external_protein_model:
+        --external_protein_model $selExtraOpts.external_protein_model
```
Quote `$selExtraOpts.external_protein_model`

nsoranzo commented on tools/raxml/raxml.xml:
```diff
@@ -0,0 +1,593 @@
+    #end if
+    ## (-q)
+    #if $selExtraOpts.multiple_model:
+        --multiple_model $selExtraOpts.multiple_model
```
Quote `$selExtraOpts.multiple_model`

nsoranzo commented on tools/raxml/raxml.xml:
```diff
@@ -0,0 +1,593 @@
+            substitution model. To specify extra options select the 'Full option list'">
+            <option value="required">Required options only</option>
+            <option value="full">Full option list</option>
+        </param>
```
Add:
```
<when value="required" />
```

nsoranzo commented on tools/raxml/raxml.xml:
```diff
@@ -0,0 +1,593 @@
+                <has_text_matching expression="Whale" />
+              </assert_contents>
+          </output>
+      </test>
```
The last 2 tests seem to have the same params.

nsoranzo commented on tools/raxml/raxml.xml:
```diff
@@ -0,0 +1,593 @@
+        --secondary_structure_model $selExtraOpts.secondary_structure_model
+    #end if
+    ## (-b)
+    #if $selExtraOpts.bootseed:
```
Here `$selExtraOpts.bootseed` is an (optional) integer, so this `if` will not pass if it is 0. You should use `str($selExtraOpts.bootseed)` instead.

nsoranzo commented on tools/raxml/raxml.xml:
```diff
@@ -0,0 +1,593 @@
+        --sliding_window_size $selExtraOpts.sliding_window_size
+    #end if
+    ## (-x)
+    #if $selExtraOpts.rapid_bootstrap_random_seed:
```
`str($selExtraOpts.rapid_bootstrap_random_seed)`

nsoranzo commented:
Thanks @tiagoantao, looks pretty good!

@martenson: I'll leave the merging and fixing to you then!

bgruening commented:
Will start working on it. @martenson is busy right now.
